go/apps/http_api_nostream/tests/test_vumi_app.py

```diff
@@ -818,12 +867,162 @@ def test_post_inbound_message_500_response(self):
+        })
+        msg_d = self.app_helper.make_dispatch_inbound(
+            'in 1', message_id='1', conv=self.conversation)
 
-        def raiser(*args, **kw):
-            raise exception_class()
-        self.patch(vumi_app, 'http_request_full', raiser)
+        # return 500 response to message push
+        req = yield self.push_calls.get()
+        req.setResponseCode(500)
+        req.finish()
+
+        # catch and check retry
+        retry = yield retry_calls.get()
+        retry_msg = self.assert_retry(retry, self.get_message_url())
+        retry.setResponseCode(200)
+        retry.finish()
+
+        with LogCatcher(log_level=logging.INFO) as lc:
+            msg = yield msg_d
+
+        self.assertEqual(lc.messages(), [
+            "Successfully scheduled retry of request [account: u'test-0-user'"
+            ", url: '%s']" % self.get_message_url(),
+        ])
+
+        self.assertEqual(retry_msg['message_id'], msg['message_id'])
+
+    @inlineCallbacks
+    def test_post_inbound_message_timeout_schedules_retry(self):
+        retry_url, retry_calls = yield self.start_retry_server()
+        yield self.start_app_worker({
+            'http_retry_api': retry_url,
+        })
+        msg_d = self.app_helper.make_dispatch_inbound(
+            'in 1', message_id='1', conv=self.conversation)
+
+        self._patch_http_request_full(HttpTimeoutError, once=True)
+
+        # catch and check retry
+        retry = yield retry_calls.get()
+        retry_msg = self.assert_retry(retry, self.get_message_url())
+        retry.setResponseCode(200)
+        retry.finish()
+
+        with LogCatcher(log_level=logging.INFO) as lc:
+            msg = yield msg_d
+
+        self.assertEqual(lc.messages(), [
+            "Successfully scheduled retry of request [account: u'test-0-user'"
+            ", url: '%s']" % self.get_message_url(),
+        ])
+
+        self.assertEqual(retry_msg['message_id'], msg['message_id'])
+
+    @inlineCallbacks
+    def test_post_inbound_message_500_retry_fails_with_500(self):
+        retry_url, retry_calls = yield self.start_retry_server()
+        yield self.start_app_worker({
+            'http_retry_api': retry_url,
+        })
+        msg_d = self.app_helper.make_dispatch_inbound(
+            'in 1', message_id='1', conv=self.conversation)
+
+        # return 500 response to message push
+        req = yield self.push_calls.get()
+        req.setResponseCode(500)
+        req.finish()
+
+        # catch and check retry
+        retry = yield retry_calls.get()
+        retry_data = retry.content.read()
+        retry.setResponseCode(500)
+        retry.finish()
+
+        with LogCatcher(log_level=logging.WARNING) as lc:
+            yield msg_d
+
+        self.assertEqual(lc.messages(), [
+            "Failed to schedule retry request [account: u'test-0-user'"
+            ", request: %r, response: [500, 'Internal Server Error']]"
+            % (retry_data,),
+        ])
```

Comment:
Based on RL discussion, the plan is to change failed retries to reraise the error so that the message doesn't get thrown away.